File: AdvancedTrack/_03-Hands On - Handling Solid Edge Events/vb/Example1/Example1/ControlExtensions.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms


Namespace Example1
	Public Module ControlExtensions
		''' <summary>
		''' Synchornous call to control but may load to deadlock if events are firing rapidly.
		''' </summary>
		<System.Runtime.CompilerServices.Extension> _
		Public Sub InvokeIfRequired(Of TControl As Control)(ByVal control As TControl, ByVal action As Action(Of TControl))
			If control.InvokeRequired Then
				control.Invoke(action, control)
			Else
				action(control)
			End If
		End Sub

		''' <summary>
		''' Asynchornous to control and should be safest.
		''' </summary>
		<System.Runtime.CompilerServices.Extension> _
		Public Sub BeginInvokeIfRequired(Of TControl As Control)(ByVal control As TControl, ByVal action As Action(Of TControl))
			If control.InvokeRequired Then
				control.BeginInvoke(action, control)
			Else
				action(control)
			End If
		End Sub
	End Module
End Namespace

File: AdvancedTrack/_03-Hands On - Handling Solid Edge Events/cs/Example1/Example1/ControlExtensions.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Windows.Forms;


namespace Example1
{
    public static class ControlExtensions
    {
        /// <summary>
        /// Synchornous call to control but may load to deadlock if events are firing rapidly.
        /// </summary>
        public static void InvokeIfRequired<TControl>(this TControl control, Action<TControl> action)
            where TControl : Control
        {
            if (control.InvokeRequired)
            {
                control.Invoke(action, control);
            }
            else
            {
                action(control);
            }
        }

        /// <summary>
        /// Asynchornous to control and should be safest.
        /// </summary>
        public static void BeginInvokeIfRequired<TControl>(this TControl control, Action<TControl> action)
            where TControl : Control
        {
            if (control.InvokeRequired)
            {
                control.BeginInvoke(action, control);
            }
            else
            {
                action(control);
            }
        }
    }
}
